src/cargo/ops/cargo_rustc/layout.rs

```diff
@@ -85,14 +85,33 @@ impl Layout {
     /// Differs from `at` in that this calculates the root path from the workspace target directory,
     /// adding the target triple and the profile (debug, release, ...).
     pub fn new(ws: &Workspace, triple: Option<&str>, dest: &str) -> CargoResult<Layout> {
+        use std::hash::{Hash, Hasher};
+        use std::collections::hash_map::DefaultHasher;
+
         let mut path = ws.target_dir();
         // Flexible target specifications often point at filenames, so interpret
         // the target triple as a Path and then just use the file stem as the
         // component for the directory name.
         if let Some(triple) = triple {
-            path.push(Path::new(triple)
-                .file_stem()
-                .ok_or_else(|| format_err!("target was empty"))?);
+            if triple.ends_with(".json") {
+                let json_path = Path::new(triple).canonicalize().map_err(|err| {
```

Comment:
This can use `chain_err` to avoid explicitly adding the string of `err` to the error message